Bilinear Upsampling increased throughput

ThisIsIsaac opened the pull request:
changed `UpsampleBilinearKernel` s.t. the throughput increased 40~50%.

I tested locally with my local test code -- **not pytorch's provided test code** -- because I am having a build problem ( which I made an issue about [here](https://github.com/pytorch/pytorch/issues/19184)). I tested with various tensor sizes and across all the sizes, it should a significant increase in throughput.

1. added `__restrict__`
2. instead of launch as many threads as there are output elements, I launched only `output_height * output_width` may threads and had each thread iterate through the channel and batch dimension.

ThisIsIsaac committed:
Update upsample_op.cu

ThisIsIsaac committed:
Update upsample_op.cu

ThisIsIsaac committed:
Update upsample_op.cu

ThisIsIsaac committed:
copied and pasted code from pytorch master

copied and pasted code from pytorch master to check if this passes the test ( since my version of the code passes my local test which tests my kernel against the copied and pasted code)

ThisIsIsaac committed:
fixed error from copying as pasting

There were some changes that weren't reverted.

ThisIsIsaac committed:
Update upsample_op.cu

ThisIsIsaac committed:
loop over batch

corrected mistake of not iterating over the batch dimension.

ThisIsIsaac commented:
@zou3519 CUDA throughput increase completed from the issue

zou3519 commented:
What are the sizes you're benchmarking with? Could you provide a benchmarking script and some numbers?

zou3519 commented on caffe2/operators/upsample_op.cu:
```diff
@@ -1,3 +1,5 @@
+// copied and pasted from pytorch to test if this passes the build.
```
Just checking -- you are modifying Caffe2's bilinear upsampling code?

ThisIsIsaac replied:
You can ignore that comment. that was because I was failing some tests but when I checked locally against the unmodified upsampling code, the results were identical. So I was wondering if the tests were faulty or the original code had errors. Turns out I wasn't iterating over the batch dimension.

Yes, I am modifying Caffe2's bilinear upsampling code.

ThisIsIsaac commented:
Attached file with test and profiling code.
[test_upsample.zip](https://github.com/pytorch/pytorch/files/3092194/test_upsample.zip)

zou3519 commented:
@ThisIsIsaac I'm confused -- I don't see a run_test.cu in the zip file:
![image](https://user-images.githubusercontent.com/5652049/56436850-49499700-62ab-11e9-8456-6df1d670f8bb.png)

ThisIsIsaac commented:
oh I'm sorry. This contains all the required files.
[test_upsample.zip](https://github.com/pytorch/pytorch/files/3098982/test_upsample.zip)

ThisIsIsaac committed:
bug fix: not iterating over all elements

when `output_height * output_width > blockDim.x * gridDim.x`, there was a bug of not iterating through all the elements.

ThisIsIsaac commented:
@zou3519 
benchmarking results:
```
batch = 1
num_channels = 100
input_height = 128
input_width = 128
width_scale = 2.f
height_scale = 2.f

// this was the last run of 10 iterations, so the initial GPU warm-up cost is not included
ref duration 266ms
my duration 173ms
```


```
batch = 10
num_channels = 100
input_height = 128
input_width = 128
width_scale = 2.f
height_scale = 2.f

// this was the last run of 10 iterations, so the initial GPU warm-up cost is not included
ref duration 2489ms
my duration 1744ms
```

```
batch = 100
num_channels = 100
input_height = 128
input_width = 128
width_scale = 2.f
height_scale = 2.f

// this was the last run of 10 iterations, so the initial GPU warm-up cost is not included
ref duration 29750ms
my duration 20549ms
```

```
batch = 1
num_channels = 1
input_height = 128
input_width = 128
width_scale = 2.f
height_scale = 2.f

// this was the last run of 10 iterations, so the initial GPU warm-up cost is not included
ref duration 16ms
my duration 17ms
```

ThisIsIsaac commented:
this doesn't seem to test `upsample_op.cu` and fails at `conv_op.cc` for some reason. could you help me figure out the error? Thanks
[build_1407584_step_103_container_0.txt](https://github.com/pytorch/pytorch/files/3101962/build_1407584_step_103_container_0.txt)

zou3519 commented:
It looks like the build timed out on the `caffe2_py2_mkl_ubuntu16_04_test`. However, if you scroll down, one of the other tests might have a test for upsample_op.cu (caffe2_py2_cuda9_1_cudnn7_ubuntu16_04_build is one such build that has cuda enabled).

I'm not very familiar with this part of the codebase so I don't know how the tests work. cc @jerryzh168, could you take a look at this please?

ThisIsIsaac commented:
@zou3519 
Thanks for the info. I will take a look in the moring.

I was looking into how `THC_pointwiseApply3` and such variants defined in [THCApply.cuh](https://github.com/pytorch/pytorch/blob/e45e3634d6dff9384d899a6a3623572342af6032/aten/src/THC/THCApply.cuh#L113) launches as many threads as the total number of elements. If I apply the same optimization strategy ( instead of mapping each thread to every element, we only spawn as many threads as width * height, and have each thread iterate through other dimensions ), we can see similar speedup while using less SMs. Let me know what you think.

Thanks as always.

zou3519 commented:
Do you have any insight into why launching as many threads as width * height is better than launching as many threads as per elements?

jerryzh168 commented:
@zou3519 are you talking about the CI failures, it looks like it fails in Too long with no output (exceeded 1h0m0s)? so I suspect it might be some tests from distributed training. we can probably ignore or if you rerun it might pass.

ThisIsIsaac commented:
@jerryzh168 
Thanks for the info. I haven't change any file other than `upsample_op.cu` and the hanging seems to occur with `conv2D`. would you suggest I just push another empty commit to rerun the tests?

ThisIsIsaac commented:
[pt_bilinear_upsample.nsight-cuprof-report.zip](https://github.com/pytorch/pytorch/files/3106136/pt_bilinear_upsample.nsight-cuprof-report.zip)
@zou3519 
I have attached two files: output file from nsight compute which you can open with nsight compute and a pdf file of the profiled data ( i.e. that is what you would see if you were to open the profiling data with nsight compute ).

I have used 2 optimization strategies, which each had different benefits to the throughput:

## looping over batch x channel
- decreased the number of computations. Under "scheduler Statistics" you can see that the numer of executed instructions per cycle decreased significantly. Also, the bar graph named "GPU Utilization" shows that the bottleneck shifted from computation to memory.

## using `const` and `__restrict__`
- under "memory workload analysis" you can see that "L1 Hit rate" went from 0 to 54.09% ( indicated by `(+inf%)` ). Also the "memory chart" shows that Unified Cache has 54.09 hit rate where as the baseline had 0%. 

Decreasing the number of computations had the biggest effect ( 20~30%) and using `__restrict__` had marginal effect (3~5%).
[upsample_profile.pdf](https://github.com/pytorch/pytorch/files/3106135/upsample_profile.pdf)

zou3519 commented:
@ThisIsIsaac thank you for the analysis, I will take a look at it later today. 

Re: THC_pointwiseApply3: because it is in our legacy codebase, ideally we would move all code using it to ATen and use TensorIterator (ATen's abstraction for applying operations over tensors). Here's an example of a [function implemented with TensorIterator]( https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/BinaryOpsKernel.cu#L42) .

However, since the change you're suggesting (launch with fewer blocks) should not be too hard to add to THC_pointwiseApply3, feel free to experiment with it. If it indeed does improve perf then that would be great.

jerryzh168 commented:
> @jerryzh168
> Thanks for the info. I haven't change any file other than `upsample_op.cu` and the hanging seems to occur with `conv2D`. would you suggest I just push another empty commit to rerun the tests?

you can comment with "@pytorchbot retest this please" to rerun the tests

ThisIsIsaac commented:
@pytorchbot retest this please

ThisIsIsaac commented:
@jerryzh168 
I'm not sure why, but the builds are still failing.

jerryzh168 commented:
doesn't look like it's relevant: `Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/caffe2/py2-mkl-ubuntu16.04:266-5c68e193bab924129bb05a2dee404f75e7458dad not found`

ThisIsIsaac commented:
@zou3519 

> doesn't look like it's relevant: Error response from daemon: manifest for 308535385114.dkr.ecr.us-east-1.amazonaws.com/caffe2/py2-mkl-ubuntu16.04:266-5c68e193bab924129bb05a2dee404f75e7458dad not found

It seems like all tests are fine. 

I would recommend adding two test cases:
1. where there are more elements in elements in`output_width x output_height` then there are spanwed threads.
2. where there are more elements in elements in`input_width x input_height` then there are spanwed threads.

My first commit had a bug of not fully iterating through all the elements when there were more elements than spanwed threads in the `output_width x output_height`  dimension, but no test case tested for this and my buggy code passed all tests.

ThisIsIsaac commented:
@zou3519 
ping

ThisIsIsaac commented:
@zou3519
Please authorize me to merge once you are done reviewing.

jerryzh168 commented:
@BIT-silence could you review this diff?

ThisIsIsaac commented:
@BIT-silence ping

BIT-silence commented on caffe2/operators/upsample_op.cu:
```diff
@@ -19,33 +21,30 @@ inline __device__ int idx(
+    float* __restrict__ Y) {
+
+
+  for (int index = threadIdx.x + blockIdx.x * blockDim.x; index < output_height * output_width; index += blockDim.x * gridDim.x){
```
```suggestion
    const int size = output_height * output_width;
    CUDA_1D_KERNEL_LOOP(index, size) {
```

BIT-silence commented on caffe2/operators/upsample_op.cu:
```diff
@@ -61,42 +60,21 @@ __global__ void UpsampleBilinearKernel(
-                      input_width,
-                      h1 + h1p,
-                      w1 + w1p)]));
+    for (int n = 0; n < num_batch; n++){
```
Not sure if here we can have some better solution. But it's fine to keep it as this now.

xiaomengy reviewed (approved):
LGTM

ThisIsIsaac committed:
Update caffe2/operators/upsample_op.cu

Co-Authored-By: Xiaomeng Yang <bit.yangxm@gmail.com>

ThisIsIsaac commented:
@BIT-silence 
Great! Thanks for taking a look.

ezyang commented:
@pytorchbot rebase this please

Commit:
Merge remote-tracking branch 'origin/master' into HEAD
